Tests/NIOWebSocketTests/WebSocketClientEndToEndTests.swift

```diff
@@ -0,0 +1,428 @@
+        }
+    }
+}
+
+extension ByteBuffer {
+    fileprivate static func forString(_ string: String) -> ByteBuffer {
+        var buf = ByteBufferAllocator().buffer(capacity: string.utf8.count)
+        buf.writeString(string)
+        return buf
+    }
+}
+
+private func setUpClientChannel(clientHTTPHandler: RemovableChannelHandler,
+                                clientUpgraders: [NIOHTTPClientProtocolUpgrader],
+                                _ upgradeCompletionHandler: @escaping (ChannelHandlerContext) -> Void) throws -> EmbeddedChannel {
+    
+    let channel = EmbeddedChannel()
+    
+    let config: NIOHTTPClientUpgradeConfiguration = (
+        upgraders: clientUpgraders,
+        completionHandler: { context in
+            channel.pipeline.removeHandler(clientHTTPHandler, promise: nil)
+            upgradeCompletionHandler(context)
+    })
+    
+    try channel.pipeline.addHTTPClientHandlers(withClientUpgrade: config).flatMap({
+        channel.pipeline.addHandler(clientHTTPHandler)
+    }).wait()
+    
+    try channel.connect(to: SocketAddress(ipAddress: "127.0.0.1", port: 0))
+        .wait()
+    
+    return channel
+}
+
+// A HTTP handler that will send an initial request which can be augmented by the upgrade handler.
+private final class BasicHTTPHandler: ChannelInboundHandler, RemovableChannelHandler {
+    fileprivate typealias InboundIn = HTTPClientResponsePart
+    fileprivate typealias OutboundOut = HTTPClientRequestPart
+    
+    fileprivate func channelActive(context: ChannelHandlerContext) {
+        // We are connected. It's time to send the message to the server to initialise the upgrade dance.
+        self.fireSendRequest(context: context)
+    }
+}
+
+// A HTTP handler that will send a request and then fail if it receives a response or an error.
+// It can be used when there is a successful upgrade as the handler should be removed by the upgrader.
+private final class ExplodingHTTPHandler: ChannelInboundHandler, RemovableChannelHandler {
+    fileprivate typealias InboundIn = HTTPClientResponsePart
+    fileprivate typealias OutboundOut = HTTPClientRequestPart
+    
+    fileprivate func channelActive(context: ChannelHandlerContext) {
+        // We are connected. It's time to send the message to the server to initialise the upgrade dance.
+        self.fireSendRequest(context: context)
+    }
+    
+    fileprivate func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        XCTFail("Received unexpected read")
+    }
+    
+    fileprivate func errorCaught(context: ChannelHandlerContext, error: Error) {
+        XCTFail("Received unexpected error")
+    }
+}
+
+extension ChannelInboundHandler where OutboundOut == HTTPClientRequestPart {
+    
+    fileprivate func fireSendRequest(context: ChannelHandlerContext) {
+        
+        var headers = HTTPHeaders()
+        headers.add(name: "Content-Type", value: "text/plain; charset=utf-8")
+        headers.add(name: "Content-Length", value: "\(0)")
```

Comment:
Hmm, this API is making me a bit uncertain now.

Given that most clients will never support upgrading to more than one protocol, I wonder if the right thing to do is to simplify `HTTPClientUpgradeHandler` to only allow one upgrader, and then to give the upgrader the ability to configure the upgrade request. That way, rather than forcing the client to send the HTTP request that will be mutated, we could instead have the upgrader emit that request, making all of this substantially more automatic. What do you think @tigerpixel?

Reply:
I originally did it in a similar way but on the upgrade handler. I don't think it is a bad idea, but I do have some comments about automatically sending the request and allowing flexibility.  I wonder if the vapor guys have comments too as they also do it this way. How do you feel about opening a separate issue to discuss it? If you would rather not do that, I can offer a more descriptive reply here.

Reply:
I think to begin with let's start here: we can always move those comments elsewhere.

Reply:
After discussion internally I think this is actually fine for now.

Reply:
Okay, great, thanks.